### ISSUE TITLE
Refactor Runner to allow Algorithm to own training loop

### ISSUE THREAD

**Issue by krzentner** [rlworkgroup/garage#704]
Currently, the [`LocalTfRunner._train`](https://github.com/rlworkgroup/garage/blob/8b1bac869ac2bb167aa9982c80ee968e662b9f5a/src/garage/experiment/local_tf_runner.py#L352) method owns the outer training loops, because it uses them to provide services (such as snapshotting), to the Algorithm. However, all of those services could be provided by a "magic iterator" over the epoch.

Concretely, this would mean that the `LocalTfRunner._train` method would look like this:
```python
last_return = None
itr = 0
for epoch in self.step_epochs():
    for cycle in range(n_epoch_cycles):
        paths = self.obtain_samples(itr, batch_size)
        last_return = self.algo.train_once(itr, paths)
        itr += 1
return last_return
```

Once `LocalTfRunner._train` looks like this (since it will then use public APIs), we can move it to `BaseAlgorithm` (and perhaps rename it). We should probably also consider changing the sampler API, so that the Algorithm has direct access to the sampler, and can pass it a policy to obtain samples with.

**Comment by naeioi** [comment]
I don't quite get the point of this proposal. Can you elaborate a bit? In particular, if we were to place the training loop in algorithm, and give an algorithm access to sampler, isn't it essentially reverting to the garage before runner came to exist? 

To me, the idea of runner owning training loop is to decouple algorithm from experiment orchestration. This includes sampling, logging, snapshotting, and the data piping between components in garage. So that algorithm cares only about updating policy given a batch of samples.

**Comment by krzentner** [comment]
For most deep RL algorithms, this change won't make a big difference, since they are easy to write using the loop described above by defining `train_once`. However, for algorithms that operate in other ways (such as a collection of policies), this is really awkward. For example, evolutionary strategies algorithms need to switch out the policy each `train_once` step, and match up the `paths` it receives with the previous `policy`. We would like it to be easy to express algorithms like that. However, we still want the runner to own all of other "services" like sampling, logging, and snapshotting.

I think that the refactoring we're doing here will let us do that, by exposing a simple interface for all of those other services. What we're giving up in the process is that the runner no longer controls data pipelining between the sampler and the algorithm, and instead the algorithm interacts with the sampler through the sampler interface. This seems sufficient to me, since I don't see what else we're likely to do, and the dataflow here is just two lines of code, so it's fine if the algorithm owns it.

**Comment by naeioi** [comment]
I think you just hit the right point on the problem of the current design. I can relate a lot to your point on revolutionary algorithm having to switch policy each `train_once` step. This happens on both CEM and CMAES, and can be problematic in many ways. I want to support this with a concrete example - for these two algorithms, there's an implicit requirement that `n_epoch_cycles` in runner should equal to `n_samples` in algo, and I failed to find an elegant way of making it explicit. 

To test if #704 really makes a difference for revolutionary alorithms, it will be great if CEM can be refactored so that in `train_once` it doesn't need to check _stage_ using `itr` and do different thing depending on stage. (See code comment in CEM for stages in this algorithm.)

**Comment by zequnyu** [comment]
@naeioi Thank you! Will try it.

**Comment by zequnyu** [comment]
Closed via #728.
